src/gradualizer_lib.erl

```diff
@@ -170,3 +177,32 @@ get_ast_children({clauses, Clauses}) ->
 get_ast_children(Node) ->
     [_Tag, _Anno | Children] = tuple_to_list(Node),
     Children.
+
+
+%% ------------------------------------------------
+%% Type environment
+%% ------------------------------------------------
+
+-spec empty_tenv() -> tenv().
+empty_tenv() ->
+    #{module => undefined,
+      types => #{},
+      records => #{}}.
+
+-spec create_tenv(_, _, _) -> tenv().
+create_tenv(Module, TypeDefs, RecordDefs) ->
+    TypeMap =
+        maps:from_list([begin
+                            Id       = {Name, length(Vars)},
+                            Params   = [VarName || {var, _, VarName} <- Vars],
+                            {Id, {Params, typelib:remove_pos(Body)}}
+                        end || {Name, Body, Vars} <- TypeDefs]),
+    RecordMap =
+        maps:from_list([{Name, [{typed_record_field, Field, typelib:remove_pos(Type)}
+                                || {typed_record_field, Field, Type}
+                                       <- lists:map(fun absform:normalize_record_field/1,
```

Comment:
We could think about what the responsibility is of the different modules `typelib`, `absform` and `gradualizer_lib`... Maybe there is none...?

Reply:
Seems like you'd prefer a tenv module... Encapsulating an abstract datatype is a natural thing to do, but in this case there isn't much to encapsulate since the type is not opaque and we access it using pattern matching. Everything in such module would need to be exported anyway.